smartsim/_core/mli/infrastructure/worker/torch_worker.py

```diff
@@ -61,65 +67,134 @@ def load_model(
+            slices.append(slice(total_samples, total_samples + num_samples))
+            total_samples = total_samples + num_samples
+
+            if res_idx == len(fetch_results) - 1:
+                # For each tensor in the last input, get remaining dimensions
+                # Assumptions: all inputs have the same number of tensors and
+                # last N-1 dimensions match across inputs for corresponding tensors
+                # thus: resulting array will be of size (num_samples, all_other_dims)
+                for item_meta in fetch_result.meta:
+                    tensor_desc: tensor_capnp.TensorDescriptor = item_meta
+                    tensor_dims = list(tensor_desc.dimensions)
+                    all_dims.append([total_samples, *tensor_dims[1:]])
+                    all_dtypes.append(str(tensor_desc.dataType))
+
+        for result_tensor_idx, (dims, dtype) in enumerate(zip(all_dims, all_dtypes)):
+            itemsize = np.empty((1), dtype=dtype).itemsize
+            alloc_size = int(np.prod(dims) * itemsize)
+            mem_alloc = mem_pool.alloc(alloc_size)
+            mem_view = mem_alloc.get_memview()
+            mem_view[:alloc_size] = b"".join(
+                [
+                    fetch_result.inputs[result_tensor_idx]
+                    for fetch_result in fetch_results
+                ]
             )
-        return TransformInputResult(result)
-        # return data # note: this fails copy test!
 
+            results.append(mem_alloc.serialize())
+
+        return TransformInputResult(results, slices, all_dims)
+
+    # pylint: disable-next=unused-argument
     @staticmethod
     def execute(
-        request: InferenceRequest,
+        batch: InferenceBatch,
         load_result: LoadModelResult,
         transform_result: TransformInputResult,
+        device: str,
     ) -> ExecuteResult:
         if not load_result.model:
             raise SmartSimError("Model must be loaded to execute")
+        device_to_torch = {"cpu": "cpu", "gpu": "cuda"}
+        for old, new in device_to_torch.items():
+            device = device.replace(old, new)
+
+        tensors = []
+        mem_allocs = []
+        for transformed, dims in zip(
+            transform_result.transformed, transform_result.dims
+        ):
+            mem_alloc = MemoryAlloc.attach(transformed)
+            mem_allocs.append(mem_alloc)
+            tensors.append(
+                torch.from_numpy(
+                    np.frombuffer(
+                        mem_alloc.get_memview()[0 : np.prod(dims) * 4], dtype=np.float32
+                    ).reshape(dims)
+                )
+            )
 
         model: torch.nn.Module = load_result.model
-        model.eval()
-        results = [model(tensor).detach() for tensor in transform_result.transformed]
+        with torch.no_grad():
+            model.eval()
+            results = [
+                model(tensor.to(device, non_blocking=True)).detach()
```

Comment:
I think I wrote this line incorrectly. It should actually just call model on a list of tensors (because a model can accept a list of tensors as input) once. I need to do a sanity check with a multi-input model.